### DISCUSSION TITLE
VS Code as AaC UI?

### DISCUSSION THREAD

**Discussion by jondavid-black** [DevOps-MBSE/AaC#171] [Q&A]
Over the holidays I've been thinking about the best way to approach the AaC UI.  One thing we did in the past was to just use Eclipse plugins for language editing support.  This worked well for developers but wasn't visual in nature.  I think using an IDE was also intimidating to some users.  That said, code editors like VS Code are pretty powerful and may be a valid approach for the AaC UI.  

I think it is fair to assume having a VS Code extension for AaC is something we want to do to support developers.  The question here:  can we use the same VS Code extension (or extension set) as the UI for other "non-developer" stakeholders?

**Comment by jondavid-black** [comment]
Some minor thoughts on use of VS Code:

1. We need to consider using a file extension we can associate with AaC.  Perhaps `myfile.aac` would be reasonable.  I've renamed a yaml file and tested validation with the `.aac` file extension and it works fine (no surprise there).
2. The VS Code example extension [Cat Scratch](https://github.com/microsoft/vscode-extension-samples/tree/main/custom-editor-sample) seems to be the closest to functionality we want to emulate with an AaC UI.
3. We'll need a good way to know valid content to support things like adding fields to a model within context.  Need to do some more digging to find a good approach.  Creating a language server may be advantageous but creates a significant dependency, so some MVP may be appropriate for initial exploration.

**Reply by Coffee2Bits** [reply]
My thoughts on your first bullet: I'm fond of the idea of using `.aac` as the file type, but it seems like most other tools that leverage yaml just leave the extension as is. I'd be worried that using a custom extension might also add a dash of confusion to users since `.yaml` is a well-known text file format whereas `.aac` could be a binary file, archived file, text file, or anything we might want it to be.

**Comment by jondavid-black** [comment]
I spent the weekend learning what not to do when writing VS Code extensions.  😄 

I believe there are a few different views for AaC that would be valuable:
1. A text editor that comprehends AaC - This could likely be a customization or extension of the [YAML extension](https://github.com/redhat-developer/vscode-yaml).  The only thing missing here is a good tree view / outline as demonstrated by the VS Code example [TreeView](https://github.com/microsoft/vscode-extension-samples/tree/main/tree-view-sample).
2. A visual editor that comprehends structural context - This could look something like a hierarchy / org chart.  Each entry is a model established based on the decomposition content in the `components` of the model.  I don't believe a complete view is needed, just the parent, siblings, and children of a model.  The open AaC file is at the middle.  Clicking on another model in the structure opens it and re-centers the view.  The central entry contains controls to add new components or delete a model.
3. A model design view - This view is focused on interactions depicted within the models' components for each behavior.  Essentially this looks like a traditional UML class diagram, but each behavior includes something like an activity diagram showing interactions among the model components to achieve the expected behavior.  More to follow on this once I get a wireframe built up for it.  
4. A visual editor - This view is an alternative to the text editor for users who aren't comfortable working in the text editor.  I envision this looking something like this [JSON View](https://github.com/pgrabovets/json-view) but values are editable fields (i.e. text input, check boxes, and selection drop downs).  List headers have an 'add' button.  Entries in the list have a 'delete' button.  Required fields have a decorator.  Invalid fields show an error indicator.
5. A way to navigate - the visual UIs should provide simple buttons that allow the user to open the AaC model in a different view.  This is pretty much the only thing I managed to get working over the weekend and will figure out how to post the code I used.  Interactions in the javascript aren't completely clear or intuitive (for me anyway).

One thing I definitely learned is making something function in a scratch html/css/js context does not translate directly into a working VS Code extension.  Using 3rd party javascript libraries was not reliable for me and I don't know why...yet.

**Comment by jondavid-black** [comment]
A potentially useful library for mainpulating YAML:  [YAML-edit](https://www.npmjs.com/package/yaml-edit)

**Comment by jondavid-black** [comment]
Some initial testing shows that [Google Charts](https://developers.google.com/chart/interactive/docs) can be used for visualization within a VS Code extension.  This is an [Org Chart](https://developers.google.com/chart/interactive/docs/gallery/orgchart) that could potentially be used for visualizing the structure within the AaC model.

**Comment by cvchaparro** [comment]
Given the similarities, I think we could use the [Ansible Language Server](https://github.com/ansible/ansible-language-server) as a spring-board for creating an AaC Language Server.

**Reply by Coffee2Bits** [reply]
I came across this [list](https://microsoft.github.io/language-server-protocol/implementors/servers/) while researching the VSCode Plugins, it's a list of language servers with the language they're written for and in. There are several written in python, like the CWL and the  Fortran Language server.

**Reply by cvchaparro** [reply]
Another similar resource: https://langserver.org/
